### PR TITLE
Browse-Path functionality

### PR DESCRIPTION
**Pull request by m-alreeni**
-  Browse-Path functionalityto check feasibility of getting XPath expression for NodeId from browsepath
- it was originally written using W3C library and then be switched directly to dom4j, so it needs improvements and optimization. 
- This funtionality has been applied at start to first aproach for extending config and should be applied for finalized approved config..

### CODE REVIEW

File: src/main/java/com/sap/dsc/aas/lib/ua/transform/BrowsepathXPathBuilder.java [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        hierarchyIsConstraint="(";
+        for (String ref : hierarchyReferences) {
+            hierarchyIsConstraint += "@ReferenceType=\"" + ref + "\" or ";
+        }
+        hierarchyIsConstraint = hierarchyIsConstraint.substring(0, hierarchyIsConstraint.length() - 3) + ") and @IsForward= \"false\"";
+    }
+
+    @Override
+    public String pathExpression(String[] browosePath) {
+        if (browosePath == null || browosePath.length == 0 || browosePath[0] == null || browosePath[0].replace("/", "").trim().isEmpty()) {
```
Comment (kenwenzel) [inline src/main/java/com/sap/dsc/aas/lib/ua/transform/BrowsepathXPathBuilder.java:50]: Please correct browosePath to browsePath.

File: src/main/java/com/sap/dsc/aas/lib/config/pojo/AbstractConfigFromAttribute.java [hunk 2]
```diff
@@ -23,7 +23,7 @@ public String getXPath() {
     @JsonProperty("from_xpath")
     public void setXPath(String xPath) {
         if (this.xPath != null) {
-            throw new AlreadyDefinedException("from_attributeName");
+            throw new AlreadyDefinedException("from_xpath_builder");
```
Comment (kenwenzel) [inline src/main/java/com/sap/dsc/aas/lib/config/pojo/AbstractConfigFromAttribute.java:26]: seems to be superfluous as from_xpath_builder is not available anymore

File: src/main/java/com/sap/dsc/aas/lib/ua/transform/BrowsepathXPathBuilder.java [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+            parent = parents.get(0);
+        } else {
+            for (int p = 1; p < parents.size(); p++) {
+                if (isParent((Element)prev, (Element)parents.get(p))) {
+                    parent = parents.get((p));
+                    break;
+                }
+            }
+        }
+        if (parent == null || (parent.getNodeType() == Node.ELEMENT_NODE && ((Element) parent).attributeCount() == 0 ))
```
Comment (kenwenzel) [inline src/main/java/com/sap/dsc/aas/lib/ua/transform/BrowsepathXPathBuilder.java:108]: use blocks with {}

File: src/main/java/com/sap/dsc/aas/lib/ua/transform/BrowsepathXPathBuilder.java [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+import org.dom4j.Node;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.invoke.MethodHandles;
+import java.util.*;
+import java.util.stream.Collectors;
+
+//FIXME the purpose is to check feasibility of getting XPath expression for NodeId from browsepath
+// it was originally written using W3C library and then be switched directly to dom4j, so it needs a lot of improvements and optimization
```
Comment (kenwenzel) [inline src/main/java/com/sap/dsc/aas/lib/ua/transform/BrowsepathXPathBuilder.java:16]: either remove this comment or document the required improvements or optimizations

**Review by kenwenzel — COMMENTED** [review]
PR can be merged after minor corrections are applied.

File: src/main/java/com/sap/dsc/aas/lib/config/pojo/AbstractConfigFromAttribute.java (hunk 2 above)
Comment (m-alreeni) [reply src/main/java/com/sap/dsc/aas/lib/config/pojo/AbstractConfigFromAttribute.java:26]: no longer needed but still remains

**Pull request merged**